base/loading.jl

```diff
@@ -34,65 +34,103 @@ function find_source_file(file)
+    return m != 0
+end
 
-    if myid() == 1 && toplevel_load
-        refs = Any[ @spawnat p _require(path) for p in filter(x->x!=1, procs()) ]
-        _require(path)
-        for r in refs; wait(r); end
+function _require_from_serialized(node::Int, path_to_try::ByteString)
+    if myid() == 1 && current_module() === Main && nprocs() > 1
+        # broadcast top-level import/using from node 1 (only)
+        if node == myid()
+            content = open(readbytes, path_to_try)
+        else
+            content = remotecall_fetch(node, open, readbytes, path_to_try)
+        end
+        if _include_from_serialized(content)
+            others = filter(x -> x != myid(), procs())
+            refs = Any[ @spawnat p _include_from_serialized(content) for p in others]
+            for (id, ref) in zip(others, refs)
+                if !fetch(ref)
+                    warn("node state is inconsistent: node $id failed to load serialized cache from :node $node:$path_to_try.")
+                end
+            end
+            return true
+        end
+    elseif node == myid()
+        if ccall(:jl_restore_incremental, UInt, (Ptr{Uint8},), path_to_try) != 0
+            return true
+        end
     else
-        _require(path)
+        content = remotecall_fetch(node, open, readbytes, path_to_try)
+        if _include_from_serialized(content)
+            return true
+        end
     end
-    nothing
+    # otherwise, continue search
+    return false
 end
 
-function _require(path)
-    global toplevel_load
-    if haskey(package_list,path)
-        loaded, c = package_locks[path]
-        !loaded && wait(c)
-    else
-        last = toplevel_load
-        toplevel_load = false
-        try
-            reload_path(path)
-        finally
-            toplevel_load = last
+function _require_from_serialized(node::Int, mod::Symbol)
+    name = string(mod)
+    finder = @spawnat node @task find_in_cache_path(mod) # TODO: switch this to an explicit Channel
+    while true
+        path_to_try = remotecall_fetch(node, finder->consume(fetch(finder)), finder)
+        path_to_try === nothing && return false
+        if _require_from_serialized(node, path_to_try)
+            return true
         end
     end
 end
 
-function reload(name::AbstractString)
-    global toplevel_load
-    path = find_in_node1_path(name)
-    path == nothing && throw(ArgumentError("$name not found in path"))
-    refs = nothing
-    if myid() == 1 && toplevel_load
-        refs = Any[ @spawnat p reload_path(path) for p in filter(x->x!=1, procs()) ]
+# to synchronize multiple tasks trying to import/using something
+package_locks = Dict{Symbol,Condition}()
+
+# require always works in Main scope and loads files from node 1
+function require(mod::Symbol)
+    loading = get(package_locks, mod, false)
+    if loading !== false
+        # load already in progress for this module
+        wait(loading)
+        return
     end
-    last = toplevel_load
-    toplevel_load = false
```

Comment:
also, fwiw, that's not the failure case that changed. the old behavior protected against the odd case of calling `require` everywhere on some code that didn't define a module, but did \*conditionally\* `require` some more code (presumably depending on `myid()`). the example you gave is just a boring no-op in both the old and new code.

i don't really see the importance of that case.

and, anyways, it's an error to call import (aka the to-be-renamed function `require`) on something that didn't define a module.